### PR TITLE
OKTA-443884 Show biometrics error in loopback polling page when device is clamshell mode

### PR DESCRIPTION
**Pull request by yannongli-okta**
## Description:

This is the SIW PR for loopback biometrics required error when device is in clamshell mode.

- Design wiki https://oktawiki.atlassian.net/wiki/spaces/~988051222/pages/2402779541/Mini+Design+Show+Error+Message+for+Loopback+in+Clamshell+mode
- Demo https://okta.box.com/s/ickvc22q36cvg4lj6jwgafkz6mvv85hs

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [x] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video:

For Desktop:

Screen Shot 2021-11-12 at 12 39 14 AM

For Mobile:

Screen Shot 2021-11-12 at 12 38 57 AM
### Reviewers:

@okta/device-platform 

### Issue:

- [OKTA-443884](https://oktainc.atlassian.net/browse/OKTA-443884)

### CODE REVIEW

File: src/v2/view-builder/internals/BaseForm.js [hunk 1]
```diff
@@ -133,6 +153,8 @@ export default Form.extend({
         if(!obj?.class || obj.class === 'INFO') {
           // add message as plain text
           this.add(`<p>${obj.message}</p>`, '.ion-messages-container');
+        } else if (obj.i18n?.key === OV_UV_ENABLE_BIOMETRIC_FASTPASS) {
```
Comment (haishengwu-okta) [inline src/v2/view-builder/internals/BaseForm.js:156]: Shall this go Identify View or Terminal View?

**Review by haishengwu-okta — CHANGES_REQUESTED** [review]

File: src/v2/view-builder/internals/BaseForm.js (hunk 1 above)
Comment (yannongli-okta) [reply src/v2/view-builder/internals/BaseForm.js:156]: @haishengwu-okta Originally i added in Terminal view, but after I rebased there is another commit that move the show messages logic in the BaseForm https://github.com/okta/okta-signin-widget/commit/63e3765952b8493b48d1386a37d3f02fc423f6c8 So i changed the way I implemented it and moved it to the BaseForm.
Shall I override showMessages in TerminalView? cc @pradeepdewda-okta

Comment (haishengwu-okta) [reply src/v2/view-builder/internals/BaseForm.js:156]: Yeah, I think Pradeep would give guideline.
IMHO `BaseForm` shall not have logic for specific flow.

File: packages/@okta/i18n/src/properties/login.properties [hunk 2]
```diff
@@ -997,7 +997,7 @@ oie.okta_verify.enroll.force.upgrade.title = Update Okta Verify
 oie.authenticator.app.non_fips_compliant_enrollment_device_incompatible = The device used to set up Okta Verify does not meet your organization’s security requirements because it is not FIPS compliant. Contact your administrator for help.
 oie.authenticator.app.non_fips_compliant_enrollment_app_update_required = The Okta Verify version on the device used does not meet your organization’s security requirements. To add your account, update Okta Verify to the latest version, then try again.
 oie.authenticator.app.method.push.verify.enable.biometrics = Your response was received, but your organization requires biometrics. Make sure your device supports biometrics, Okta Verify is up-to-date and biometrics are enabled for your account in Okta Verify, then try again.
-oie.authenticator.app.method.push.verify.enable.biometrics.title = Enable biometrics in Okta Verify
+oie.authenticator.app.method.push.verify.enable.biometrics.title = Biometrics needed for Okta Verify
```
Comment (pradeepdewda-okta) [inline packages/@okta/i18n/src/properties/login.properties:1000]: We should a message that also clarify what needs to be done ?

**Review by pradeepdewda-okta — COMMENTED** [review]
add screenshot or video.

File: src/v2/view-builder/internals/BaseForm.js (hunk 1 above)
Comment (pradeepdewda-okta) [reply src/v2/view-builder/internals/BaseForm.js:156]: Haisheng is right.. This should not be included in BaseForm.

This should go to the relevant view that renders this flow..

File: test/testcafe/spec/TerminalView_spec.js [hunk 3]
```diff
@@ -173,3 +178,20 @@ async function setupOtpOnly(t) {
               
     });
 });
+
+[  
+  ['should have biometrics required error in loopback polling page', terminalBiometricsReqruiedForFastPassMock],
```
Comment (pradeepdewda-okta) [inline test/testcafe/spec/TerminalView_spec.js:183]: Add the test in it's own view specific testcafe file.. Do not add in TerminalView

File: src/v2/view-builder/internals/BaseForm.js (hunk 1 above)
Comment (pradeepdewda-okta) [reply src/v2/view-builder/internals/BaseForm.js:156]: basically you can override `showMessages` in your respective views to display custom terminal messages..

**Comment by pradeepdewda-okta** [comment]
@yannongli-okta add a video or screenshot.

File: test/testcafe/spec/TerminalView_spec.js [hunk 4]
```diff
@@ -172,4 +172,4 @@ async function setupOtpOnly(t) {
       }
               
     });
-});
+});
```
Comment (pradeepdewda-okta) [inline test/testcafe/spec/TerminalView_spec.js:175]: nit

**Comment by yannongli-okta** [comment]
> @yannongli-okta add a video or screenshot.

@pradeepdewda-okta sure. Thanks for the reminder. I will be recording e2e demo with local monolith, so it takes a while to upload. Will update it in PR once done.

**Review by pradeepdewda-okta — APPROVED** [review]

**Comment by jhe-okta** [comment]
Is the change needed for the identifier FastPass flow? Also, the callout box's bottom border does not line up with the red box border on the left. Is this fixable?

**Comment by pradeepdewda-okta** [comment]
> Is the change needed for the identifier FastPass flow (DeviceChallengePollView)? Also, the callout box's bottom border does not line up with the red box border on the left. Is this fixable?

Good catch. This must be something recent, might have caused this css issue.

**Review by stevennguyen-okta — APPROVED** [review]

**Comment by yannongli-okta** [comment]
> Is the change needed for the identifier FastPass flow (DeviceChallengePollView)? Also, the callout box's bottom border does not line up with the red box border on the left. Is this fixable?

This change does not apply on identifier view. @jhe-okta

**Review by jhe-okta — APPROVED** [review]
The misaligned callout box border does not seem to be an implementation issue in this change. Can you please file a JIRA to follow up?

**Comment by yannongli-okta** [comment]
created follow up Jira for border not aligned error https://oktainc.atlassian.net/browse/OKTA-447252

**Comment by pradeepdewda-okta** [comment]
> created follow up Jira for border not aligned error https://oktainc.atlassian.net/browse/OKTA-447252

Is it a courage issue ? Possible to debug a little bit and assign it to correct team who last changed this..

**Comment by yannongli-okta** [comment]
@pradeepdewda-okta I pinged in ui code channel. Tested the callout in the courage playground and didn't see the issue. I was advised to find if anyone changes SIW to override the error call out css. I will ping in widget channel later.

**Comment by pradeepdewda-okta** [comment]
> @pradeepdewda-okta I pinged in ui code channel. Tested the callout in the courage playground and didn't see the issue. I was advised to find if anyone changes SIW to override the error call out css. I will ping in widget channel later.

Thanks.. because if it's not courage issue, `uicore` will be responsible to make that change and this ticket that you created will remain pending without any actions. So it's important to reach to related team.

**Pull request closed without merging**